util/lua/lua.go

```diff
@@ -0,0 +1,171 @@
+	} {
+		if err := l.CallByParam(lua.P{
+			Fn:      l.NewFunction(pair.f),
+			NRet:    0,
+			Protect: true,
+		}, lua.LString(pair.n)); err != nil {
+			panic(err)
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+	l.SetContext(ctx)
+	objectValue := decodeValue(l, obj.Object)
+	l.SetGlobal("obj", objectValue)
+	err := l.DoString(script)
+	return l, err
+}
+
+// ExecuteHealthLua runs the lua script to generate the health status of a resource
+func (vm VM) ExecuteHealthLua(obj *unstructured.Unstructured, script string) (*appv1.HealthStatus, error) {
+	l, err := vm.runLua(obj, script)
+	if err != nil {
+		return nil, err
+	}
+	returnValue := l.Get(-1)
+	if returnValue.Type() == lua.LTTable {
+		jsonBytes, err := luajson.Encode(returnValue)
+		if err != nil {
+			return nil, err
+		}
+		healthStatus := &appv1.HealthStatus{}
+		err = json.Unmarshal(jsonBytes, healthStatus)
+		if err != nil {
+			return nil, err
+		}
+		if !appv1.IsHealthStatusCode(healthStatus.Status) {
+			return &appv1.HealthStatus{
+				Status:  appv1.HealthStatusDegraded,
+				Message: invalidHealthStatus,
+			}, nil
+		}
+
+		return healthStatus, nil
+	}
+	return nil, fmt.Errorf(incorrectReturnType, returnValue.Type().String())
+}
+
+// GetScript attempts to read lua script from config and then filesystem for that resource
+func (vm VM) GetHealthScript(obj *unstructured.Unstructured) (string, error) {
+	key := getConfigMapKey(obj)
+	if script, ok := vm.ResourceOverrides[key]; ok && script.HealthLua != "" {
+		return script.HealthLua, nil
+	}
+	return vm.getPredefinedLuaScripts(key, healthScript)
+}
+
+func getConfigMapKey(obj *unstructured.Unstructured) string {
+	gvk := obj.GroupVersionKind()
+	if gvk.Group == "" {
+		return strings.ToLower(gvk.Kind)
+	}
+	return strings.ToLower(fmt.Sprintf("%s/%s", gvk.Group, gvk.Kind))
+
+}
+
+func (vm VM) getPredefinedLuaScripts(objKey string, scriptType string) (string, error) {
+	data, err := box.MustBytes(filepath.Join(objKey, scriptType))
+	if err != nil {
+		if os.IsNotExist(err) {
+			log.Debugf("No Lua Script found for resource key '%s'", objKey)
+			return "", nil
+		}
+		return "", err
+	}
+	return string(data), nil
+}
+
+// Took logic from the link below and added the int, int32, and int64 types since the value would have type int64
+// while actually running in the controller and it was not reproducible through testing.
+// https://github.com/layeh/gopher-json/blob/97fed8db84274c421dbfffbb28ec859901556b97/json.go#L154
+func decodeValue(L *lua.LState, value interface{}) lua.LValue {
+	switch converted := value.(type) {
+	case bool:
+		return lua.LBool(converted)
+	case float64:
+		return lua.LNumber(converted)
+	case string:
+		return lua.LString(converted)
+	case json.Number:
+		return lua.LString(converted)
+	case int:
+		return lua.LNumber(converted)
+	case int32:
+		return lua.LNumber(converted)
+	case int64:
+		return lua.LNumber(converted)
```

Comment:
This style nit can't be applied because lua.LNumber is a type alias for a float64 and the style nit makes converted an interface instead of a concrete type causing the type assertion to fail.

Reply:
got it. thanks